## **Let us understand what SVM is: SVM (Support Vector Machine) is a Supervised Learning Algorithm used for regression and classification but mostly it is used for Classification.

## Importing the Dependencies

In [2]:
import numpy as np
import pandas as pd

## DATA COLLECTION AND ANALYSIS

In [4]:
#Loading the diabetes dataset to pandas Dataframe
df = pd.read_csv("diabetes.csv")

In [5]:
#print the first 5 rows of dataset
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
0            6      148             72             35        0  33.6  \
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

###### Pregnancies: Number of times pregnant

###### Glucose: Plasma glucose concentration a 2 hours in an oral glucose tolerance test

###### BloodPressure: Diastolic blood pressure (mm Hg)

###### SkinThickness: Triceps skin fold thickness (mm)

###### Insulin: 2-Hour serum insulin (mu U/ml)

###### BMI: Body mass index (weight in kg/(height in m)^2)

###### DiabetesPedigreeFunction: Diabetes pedigree function

###### Age: Age (years)

###### Outcome: Class variable (0 or 1)

In [6]:
#the number of rows and columns of dataset
df.shape

(768, 9)

##### So here we have 768 rows and 9 columns .
##### 768 rows means we have 768 examples means the data is taken from 768 persons and 9 cloumns represents the features.

In [7]:
# getting the statstical measure of data
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin   
count   768.000000  768.000000     768.000000     768.000000  768.000000  \
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [8]:
#so now lets how many cases are thier for diabetic and non diabetic
df['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

In [9]:
#mean value
df.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin   
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000  \
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

#### So here the mean value of diabetic persons are more so their is more glucose content in diabetic persons.From age also we can see that persons with more age is diabetic.

In [10]:
#separete the data and labels
X=df.drop(columns = 'Outcome',axis=1)
y = df['Outcome']


In [11]:
print(X)

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
0              6      148             72             35        0  33.6  \
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [12]:
print(y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


## Data Standarlization

#### we are doing this because as thier is difference in range value for different columns so it become difficult for machine to predict.

In [15]:
from sklearn.preprocessing import StandardScaler

In [16]:
scaler = StandardScaler()

In [17]:
# now we need to fill the data in scaler
scaler.fit(X)

StandardScaler()

In [18]:
standardized_data=scaler.transform(X)
# here we are what we are doing is we are fitting all these incosintent data with our standerd scaler function and based on that standerd scaler we are tranforming the dat to an comman range

In [19]:
print(standardized_data)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [20]:
X = standardized_data
y = df['Outcome']
# so here we will use this x and y for training our model x represents data and y represents model

In [21]:
print(X)
print(y)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]
0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


In [22]:
#split the data train and test
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,stratify=y,random_state=2)
# here we taken four variable so this x data will split into 2 arrays so firts one is x_train and second one si x_test so we train our machine learning model with this and once the model is trained we will try to evalutae our model with test date

In [23]:
# check the shape
print(X.shape, X_train.shape ,X_test.shape)

(768, 8) (614, 8) (154, 8)


##### So here we see their totally 768 examples in original dataset out of which 614 are going to be used for training data and 154 for test data

## Training the Model

In [24]:
from sklearn import svm
# we are using svm function to load our support vector machine

In [25]:
classifier = svm.SVC(kernel = 'linear')
# this will load svm model to classifier

In [26]:
# training the support vector Machine Classsifier
classifier.fit(X_train,y_train)


SVC(kernel='linear')

## Model Evaluation

##### means to check how many times our model is predicting correctly.

In [27]:
# accuracy score on the training data
X_train_prediction = classifier.predict(X_train)
from sklearn.metrics import accuracy_score
training_data_accuracy = accuracy_score(X_train_prediction , y_train)
# here all the predction made by model are stored in X_train_prediction and now we are comparing the prediction of model which is stored in x_train_predction with original lable y_train which give accuracy score of model

In [28]:
print('Accuracy score of the training data: ' ,training_data_accuracy)

Accuracy score of the training data:  0.7866449511400652


In [29]:
# so here the accuracy score is almost 78 percent so its pretty good it means out the 100 prediction our model is predicting 78 times the correct prediction

In [30]:
# accuracy score of test data
X_test_prediction = classifier.predict(X_test)

test_data_accuracy = accuracy_score(X_test_prediction , y_test)

In [31]:
print('Accuracy score of the test data: ',test_data_accuracy)

Accuracy score of the test data:  0.7727272727272727


In [32]:
# here accuracy score is 77 percent so its good evidence that model has not overtrained.

## Making Predictive System

In [33]:
# now we need to make predictive system that can predict wither person has diabetic or not.


In [34]:
input_data = (1,89,66,23,94,28.1,0.167,21)

# changing input data to numpy array
input_data_as_numpy_array = np.asarray(input_data)

#reshape the array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

# standardize the input data
std_data = scaler.transform(input_data_reshaped)
print(std_data)


prediction = classifier.predict(std_data)
print(prediction)

if(prediction[0] ==0):
  print("The person is not diabetic")
else:
  print("The person is  diabetic")  

[[-0.84488505 -0.99820778 -0.16054575  0.15453319  0.12330164 -0.49404308
  -0.92076261 -1.04154944]]
[0]
The person is not diabetic


c:\Users\ashwi\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
